To add single and chained learning schedulers to docs (#56705)
pytorch/pytorch: docs/source/optim.rst

Example::

    scheduler = ExponentialLR(optimizer, gamma=0.9)
    for epoch in range(20):
        for input, target in dataset:
            optimizer.zero_grad()
            output = model(input)
            loss = loss_fn(output, target)
            loss.backward()
            optimizer.step()
            scheduler.step()

One interesting feature of schedulers are that they can be called back-to-back and to
get chained. This will lead to by schedulers are getting applied simultaneously.

Example::

    scheduler1 = ExponentialLR(optimizer, gamma=0.9)
    scheduler2 = MultiStepLR(optimizer, milestones=[30,80], gamma=0.1)
    for epoch in range(20):
        for input, target in dataset:
            optimizer.zero_grad()
            output = model(input)
            loss = loss_fn(output, target)
            loss.backward()
            optimizer.step()
            scheduler1.step()
            scheduler2.step()

In many places during this tutorial, we will use the following template to refer to schedulers
algorithms.

    >>> scheduler = ...
    >>> for epoch in range(100):
    >>>     train(...)
    >>>     validate(...)
    >>>     scheduler.step()

> 

Reviewer: nit: does this change the formatting?

Author: actually this empty line was unnecessary, so deleted it :)

Learning rate scheduling should be applied after optimizer's update; e.g., you
should write your code this way:

Example::

    model = [Parameter(torch.randn(2, 2, requires_grad=True))]
    optimizer = SGD(model, 0.1)
    scheduler = ExponentialLR(optimizer, gamma=0.9)

    for epoch in range(20):
        optimizer.step()

>         scheduler.step()

Reviewer: Observations:
- It's easier to compare different sections of the documentation if they are as similar as possible.
- It's important to tell users where the rest of the training code would go relative to the scheduler step.
- The optimizer is stepped at every dataset iteration, not just epoch.

This is a little verbose but bridges the two types of examples presented on the documentation page.
```
    model = [Parameter(torch.randn(2, 2, requires_grad=True))]
    optimizer = SGD(model, 0.1)
    scheduler = ExponentialLR(optimizer, gamma=0.9)

    for epoch in range(20):
        for input, target in dataset:
            optimizer.zero_grad()
            output = model(input)
            loss = loss_fn(output, target)
            loss.backward()
            optimizer.step()
        scheduler.step()
```

Author: Yes, indeed this looks much more consistent with the rest of the documentation :)

To preserve consistency even more i modified chain example as well accordingly. Hope it was a right decision :)

Learning rate scheduling should be applied after optimizer's update; e.g., you
should write your code this way:

> Example::

Reviewer: good idea to adjust the formatting and be more consistent :)

    scheduler = ExponentialLR(optimizer, gamma=0.9)
    for epoch in range(20):
        for input, target in dataset:
            optimizer.zero_grad()
            output = model(input)
            loss = loss_fn(output, target)
            loss.backward()
            optimizer.step()

>             scheduler.step()

Reviewer: thanks for adding this code example! i like that it emphasize that scheduler step is done after the optimizer step.

Reviewer: nit: though some schedulers are applied only at every epoch. do you have a way of suggesting this without adding too much here? if it's cumbersome, we don't need to make that point :)

Author: Thanks for the amazing review :) and i did a little simplification. Hope you're liking the current version.

One interesting feature of schedulers are that they can be called back-to-back and to

> get chained. This will lead to by schedulers are getting applied simultaneously.

Reviewer: nit: I'd suggest removing "interesting feature" and highlighting that "most schedulers" support that feature (but not all). "simultaneously" also seems unclear to me: does it mean one override the other? I suggest:

> Most learning rate schedulers can be called back-to-back (also referred to as chaining schedulers). The result is that each scheduler is applied one after the other on the learning rate obtained by the one preceding it.

Example::

    scheduler1 = ExponentialLR(optimizer, gamma=0.9)
    scheduler2 = MultiStepLR(optimizer, milestones=[30,80], gamma=0.1)
    for epoch in range(20):
        for input, target in dataset:
            optimizer.zero_grad()
            output = model(input)
            loss = loss_fn(output, target)
            loss.backward()
            optimizer.step()
            scheduler1.step()
            scheduler2.step()

> In many places during this tutorial, we will use the following template to refer to schedulers

Reviewer: nit: this will be part of the documentation, and not a tutorial :)
```suggestion
In many places in the documentation, we will use the following template to refer to schedulers
```
